File: ruby/5kyu/scramblies.rb
# Complete the function scramble(str1, str2) that returns true if a portion of str1 characters can be rearranged to match str2, otherwise returns false.

# Notes:

#     Only lower case letters will be used (a-z). No punctuation or digits will be included.
#     Performance needs to be considered

# Input strings s1 and s2 are null terminated.

# Examples

# scramble('rkqodlw', 'world') ==> True
# scramble('cedewaraaossoqqyt', 'codewars') ==> True
# scramble('katas', 'steak') ==> False

def scramble(s1,s2)
  s2.chars.uniq.all?{|x| s2.count(x)<=s1.count(x)}
end

# p scramble('rkqodlw','world')
# true

# p scramble('cedewaraaossoqqyt','codewars')
# true

# p scramble('katas','steak')
# false

# p scramble('scriptjava','javascript')
# true

# p scramble('scriptingjava','javascript')
# true

File: javascript/5kyu/scramblies.js
// Complete the function scramble(str1, str2) that returns true if a portion of str1 
// characters can be rearranged to match str2, otherwise returns false.

// Notes:

// Only lower case letters will be used (a-z). No punctuation or digits will be included.
// Performance needs to be considered

// Input strings s1 and s2 are null terminated.

function countLetters(string) {
  return string.split('').reduce((count, letter) => {
    count[letter] ? count[letter]++ : count[letter] = 1;
    return count;
  }, {});
}

function scramble(str1, str2) {
  const firstLetterCount = countLetters(str1);
  const secondLetterCount = countLetters(str2);

  const compareLetterCounts = Object.entries(secondLetterCount).map(keyValue => {
    const key = keyValue[0];
    const value = keyValue[1];

    return firstLetterCount[key] >= value ? true : false;
  })
  
  return compareLetterCounts.every(letterFound => letterFound);
}

console.log(scramble('rkqodlw', 'world'));
// True

console.log(scramble('cedewaraaossoqqyt', 'codewars'));
// True

console.log(scramble('katas', 'steak'));
// False
